File: app/src/main/java/com/example/network_termproject/ChatRoom.java
package com.example.network_termproject;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Rect;
import android.graphics.drawable.BitmapDrawable;
import android.os.Bundle;
import android.util.Base64;
import android.view.View;
import android.view.ViewTreeObserver;
import android.view.inputmethod.InputMethodManager;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.example.network_termproject.databinding.ChatRoomLayoutBinding;
import com.example.network_termproject.network.Client;
import com.example.network_termproject.network.NetData;
import com.example.network_termproject.recycler.ChatAdapter;

import java.io.ByteArrayOutputStream;
import java.util.ArrayList;

public class ChatRoom extends AppCompatActivity {

    /*
    채팅방이 가져야할 정보들
    1. 사용자 그룹 (1:1) or 그룹 -> arraylist쓰자
    3. 카톡 내용들 -> 내부 저장소에 캐시 형태로 저장하면 될 것 같다.
     */
    private String talkDataPath;
    private String roomName;
    private ChatRoomLayoutBinding binding;
    private ChatAdapter chatAdapter;
    private ArrayList<NetData> datas;
    private NetData.Builder dataBuilder;
    private ChatRoomInfo chatRoomInfo;
    private InputMethodManager imm;
    int rootHeight = -1;
    private boolean isEmojiSelected = false;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ChatRoomLayoutBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        imm = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);


        binding.chatRoomRootView.getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
            @Override
            public void onGlobalLayout() {
                if (rootHeight == -1) {
                    rootHeight = binding.chatRoomRootView.getHeight();
                }
                Rect visibleFrameSize = new Rect();
                binding.chatRoomRootView.getWindowVisibleDisplayFrame(visibleFrameSize);
                int heightExceptKeyboard = visibleFrameSize.bottom - visibleFrameSize.top;
                if (heightExceptKeyboard < rootHeight) {
                    int keyboardHeight = rootHeight - heightExceptKeyboard;
                    binding.emojiContainer.setMaxHeight(keyboardHeight);
                }
            }
        });

        Bundle bundle = getIntent().getBundleExtra("chatRoomInfo");
        assert bundle != null;
        chatRoomInfo = (ChatRoomInfo) bundle.getSerializable("chatRoomInfo");

        init();

        binding.chatRoomToolbar.setTitle(chatRoomInfo.getRoom_id());

        binding.icon1.setOnClickListener(view -> {
            ByteArrayOutputStream baos = new ByteArrayOutputStream();
            Bitmap bitmap = ((BitmapDrawable) binding.icon1.getDrawable()).getBitmap();
            bitmap.compress(Bitmap.CompressFormat.PNG, 95, baos);
            String base64Data = Base64.encodeToString(baos.toByteArray(), Base64.DEFAULT);
            dataBuilder.setType("image").setContent(base64Data);
            isEmojiSelected = true;
        });

        chatAdapter = new ChatAdapter(datas, Client.getInstance().getName());
        binding.chatRoomRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        binding.chatRoomRecyclerView.setAdapter(chatAdapter);
        binding.chatRoomSendButton.setOnClickListener((v) -> {

            //TODO 테스트용임 이모티콘 (이미지)보내기
            if (isEmojiSelected) {
                NetData clientData = dataBuilder
                        .setName(Client.getInstance().getName())
                        .setUserId(Client.getInstance().getId())
                        .setRoomId(chatRoomInfo.getRoom_id())
                        .build();
                datas.add(clientData);
                Client.getInstance().write(clientData);
                chatAdapter.notifyDataSetChanged();
                isEmojiSelected = false;
                binding.emojiContainer.setVisibility(View.INVISIBLE);
                binding.emojiContainer.setFocusable(false);
            } else {
                NetData clientData = dataBuilder.setType("text")
                        .setName(Client.getInstance().getName())
                        .setUserId(Client.getInstance().getId())
                        .setRoomId(chatRoomInfo.getRoom_id())
                        .setContent(binding.chatRoomEditText.getText().toString())
                        .build();
                datas.add(clientData);
                Client.getInstance().write(clientData);
                chatAdapter.notifyDataSetChanged();
                binding.chatRoomEditText.setText("");
            }

        });


        binding.chatRoomEmojiButton.setOnClickListener(view -> {
            if (imm.isAcceptingText()) {
                hideKeyboard();
            }
            if (binding.emojiContainer.getVisibility() == View.INVISIBLE) {
                binding.emojiContainer.setVisibility(View.VISIBLE);
                binding.emojiContainer.setFocusable(true);
            } else {
                binding.emojiContainer.setVisibility(View.INVISIBLE);
                binding.emojiContainer.setFocusable(false);
            }
        });

        chatAdapter.notifyDataSetChanged();

    }

    private void hideKeyboard() {
        imm.toggleSoftInput(InputMethodManager.SHOW_FORCED, 0);
    }

    private void init() {
        datas = new ArrayList<>();
        Client.getInstance().setDisplay(this::display);
        Client.getInstance().setName("testName");
        dataBuilder = new NetData.Builder();
    }


    private void display(NetData data) {
        runOnUiThread(() -> {
            datas.add(data);
            chatAdapter.notifyDataSetChanged();
        });
    }

    public void getTalkData() {

    }

    public String getRoomName() {
        return roomName;
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        Client.getInstance().disconnect();
    }
}

File: app/src/main/java/com/example/network_termproject/ChatRoom.kt
package com.example.network_termproject

import android.content.Intent
import android.graphics.Bitmap
import android.graphics.ImageDecoder
import android.graphics.drawable.BitmapDrawable
import android.os.Build
import android.os.Bundle
import android.util.Base64
import android.util.Log
import android.view.View
import android.view.inputmethod.InputMethodManager
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.network_termproject.network.Client
import com.example.network_termproject.network.NetData
import com.example.network_termproject.recycler.ChatAdapter
import kotlinx.android.synthetic.main.chat_room_layout.*
import java.io.*
import java.nio.ByteBuffer
import java.util.*
import java.util.function.Consumer


@RequiresApi(Build.VERSION_CODES.P)
class ChatRoom : AppCompatActivity() {

    private val galleryRequestCode = 30
    private var chatAdapter: ChatAdapter? = null
    private var datas: ArrayList<NetData>? = null
    private var dataBuilder: NetData.Builder? = null
    private var chatRoomInfo: ChatRoomInfo? = null
    private var imm: InputMethodManager? = null
    private var isEmojiSelected = false
    private val saver = TalkDataSaver()
    private var talkSaveFile: File? = null
    private lateinit var chatDistributor: ChatDistributor

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.chat_room_layout)

        init()

        imm = getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager

        val bundle = intent.getBundleExtra("chatRoomInfo")
        chatRoomInfo = bundle?.getSerializable("chatRoomInfo") as ChatRoomInfo?

        chatDistributor = ChatDistributor()
        chatDistributor.apply {
            init(getOutputDirectory(), this@ChatRoom::display)
            updateCurrentChatRoom(chatRoomInfo!!.room_id!!)
        }
        
//        Client.instance.setDisplay(chatDistributor::distribute)

        Client.instance.chatDistributor = this@ChatRoom.chatDistributor

        talkSaveFile = File(getOutputDirectory(), "${chatRoomInfo!!.room_id!!}1.txt")
        if (talkSaveFile!!.exists()) {
            val reader = FileReader(talkSaveFile)
            val bufferedReader = BufferedReader(reader)
            val iterator = bufferedReader.lineSequence().iterator()
            while (iterator.hasNext()) {
                val saveData = saver.parseToNetData(iterator.next())
                if (Objects.nonNull(saveData)){
                    datas!!.add(saveData!!)
                }
            }
            bufferedReader.close()
        }

        setButtonEvent()

        chat_room_toolbar.title = chatRoomInfo!!.room_id

        chatAdapter = ChatAdapter(datas!!, Client.instance.name, supportFragmentManager)
        chat_room_recyclerView.apply {
            layoutManager = LinearLayoutManager(this@ChatRoom)
            adapter = chatAdapter
        }
        chatAdapter!!.notifyDataSetChanged()
    }

    private fun setButtonEvent() {
        icon_2.setOnClickListener {
            val baos = ByteArrayOutputStream()
            val bitmap = (icon_2.drawable as BitmapDrawable).bitmap
            bitmap.compress(Bitmap.CompressFormat.PNG, 95, baos)
            val imageData = Base64.encodeToString(baos.toByteArray(), Base64.DEFAULT)
            dataBuilder!!.setType("image").setContent(imageData)
            try {
                baos.close()
            } catch (e: IOException) {
                e.printStackTrace()
            }
            isEmojiSelected = true
        }
        chat_room_sendButton.setOnClickListener {

            if (isEmojiSelected) {
                val clientData = dataBuilder!!
                        .setName(Client.instance.name)
                        .setUserId(Client.instance.id!!)
                        .setRoomId(chatRoomInfo!!.room_id!!)
                        .build()
                datas!!.add(clientData)
                val header = ByteBuffer.allocate(6)
                header.apply {
                    putChar('s')
                    putInt(clientData.data.toString().toByteArray().size)
                    flip()
                }

                var buffer = ByteBuffer.allocate(6 + clientData.data.toString().toByteArray().size)
                buffer.apply {
                    put(header)
                    put(clientData.data.toString().toByteArray())
                }
                buffer.flip()

                Client.instance.write(buffer)

                chatAdapter!!.notifyDataSetChanged()
                isEmojiSelected = false
                emoji_container.apply {
                    visibility = View.INVISIBLE
                    isFocusable = false
                }
            } else {
                val clientData = dataBuilder!!.setType("text")
                        .setName(Client.instance.name)
                        .setUserId(Client.instance.id!!)
                        .setRoomId(chatRoomInfo!!.room_id!!)
                        .setContent(chat_room_editText.text.toString())
                        .build()
                datas!!.add(clientData)
                val header = ByteBuffer.allocate(6)
                header.apply {
                    putChar('s')
                    putInt(clientData.data.toString().toByteArray().size)
                    flip()
                }
                var buffer = ByteBuffer.allocate(6 + clientData.data.toString().toByteArray().size)
                buffer.apply {
                    put(header)
                    put(clientData.data.toString().toByteArray())
                }
                buffer.flip()

                Client.instance.write(buffer)
                chatAdapter!!.notifyDataSetChanged()
                chat_room_editText.setText("")
            }
        }
        chat_room_emoji_button.setOnClickListener {
            if (imm!!.isAcceptingText) {
                hideKeyboard()
            }
            if (emoji_container.visibility == View.INVISIBLE) {
                emoji_container.apply {
                    visibility = View.VISIBLE
                    isFocusable = true
                }
            } else {
                emoji_container.apply {
                    visibility = View.INVISIBLE
                    isFocusable = false
                }
            }
        }
        chat_room_image_button.setOnClickListener {
            getImageFromGallery()
        }
    }

    private fun hideKeyboard() {
        imm!!.toggleSoftInput(InputMethodManager.SHOW_FORCED, 0)
    }

    private fun init() {
        datas = ArrayList()
        dataBuilder = NetData.Builder()
    }

    private fun display(data: NetData) {
        runOnUiThread {
            datas!!.add(data)
            chatAdapter!!.notifyDataSetChanged()
        }
    }

    override fun onBackPressed() {
        val fileWriter = FileWriter(talkSaveFile, true)
        val bufferedWriter = BufferedWriter(fileWriter)
        for (data in datas!!.iterator()) {
            if(data.getType() != "image"){
                bufferedWriter.write(saver.generate(data.getName(), data.getUserId(), data.getType(), data.getContent()))
                bufferedWriter.newLine()
            }
        }

        bufferedWriter.apply {
            flush()
            close()
        }

        Client.instance.chatDistributor!!.updateCurrentChatRoom(null)
        finish()
    }

    private fun getOutputDirectory(): File {
        val mediaDir = externalMediaDirs.firstOrNull()?.let {
            File(it, resources.getString(R.string.app_name)).apply { mkdirs() }
        }
        return if (mediaDir != null && mediaDir.exists())
            mediaDir else filesDir
    }

    private fun getImageFromGallery() {
        val intent = Intent().apply {
            action = Intent.ACTION_GET_CONTENT
            type = "image/*"
        }
        startActivityForResult(intent, galleryRequestCode)
    }


    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == galleryRequestCode && resultCode == RESULT_OK && Objects.nonNull(data)) {
            val baos = ByteArrayOutputStream()
            val imageBitmap = ImageDecoder.createSource(this.contentResolver, data!!.data!!).let { ImageDecoder.decodeBitmap(it) }
            imageBitmap.compress(Bitmap.CompressFormat.PNG, 95, baos)
            val imageData = Base64.encodeToString(baos.toByteArray(), Base64.DEFAULT)
            val clientData = dataBuilder!!.setName(Client.instance.name)
                    .setType("image")
                    .setContent(imageData)
                    .setRoomId(chatRoomInfo!!.room_id!!)
                    .setUserId(Client.instance.id!!)
                    .build()

            datas!!.add(clientData)
            baos.close()

            val header = ByteBuffer.allocate(6)
            header.apply {
                putChar('s')
                putInt(clientData.data.toString().toByteArray().size)
                flip()
            }

            var buffer = ByteBuffer.allocate(6 + clientData.data.toString().toByteArray().size)
            buffer.apply {
                put(header)
                put(clientData.data.toString().toByteArray())
            }
            buffer.flip()

            Client.instance.write(buffer)

            chatAdapter!!.notifyDataSetChanged()
        }
    }
}